cpp/comms/src/ucp_helper.h

```diff
@@ -0,0 +1,157 @@
+  context->completed = 1;
+
+  printf("[0x%x] send handler called with status %d (%s)\n",
+         (unsigned int)pthread_self(), status, ucs_status_string(status));
+}
+
+/**
+ * @brief Asynchronous recv callback sets request to completed
+ */
+static void recv_handle(void *request, ucs_status_t status,
+                        ucp_tag_recv_info_t *info) {
+  struct ucx_context *context = (struct ucx_context *)request;
+  context->completed = 1;
+
+  printf("[0x%x] receive handler called with status %d (%s), length %lu\n",
+         (unsigned int)pthread_self(), status, ucs_status_string(status),
+         info->length);
+}
+
+/**
+ * @brief Asynchronously send data to the given endpoint using the given tag
+ */
+struct ucx_context *ucp_isend(ucp_ep_h ep_ptr, const void *buf, int size,
+                              int tag) {
+  struct ucx_context *ucp_request = 0;
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
+
+  ucp_request = (struct ucx_context *)ucp_tag_send_nb(
+    ep_ptr, buf, size, ucp_dt_make_contig(1), ucp_tag, send_handle);
+
+  /**
+   * On error, close endpoint, set request to completed and close
+   * @TODO: Should the request be closed here as well? How should we handle this?
+   * @TODO: Should the Python layer be attempting to reconnect endpoints when they are closed?
+   */
+  if (UCS_PTR_IS_ERR(ucp_request)) {
+    printf("unable to send UCX data message\n");
+    ucp_ep_close_nb(ep_ptr, UCP_EP_CLOSE_MODE_FLUSH);
+    return nullptr;
+  /**
+   * If the request didn't fail, but it's not OK, it is in flight.
+   * Expect the handler to be invoked
+   */
+  } else if (UCS_PTR_STATUS(ucp_request) != UCS_OK) {
+    printf("Message is sending. Handler should be invoked.\n");
+
+   /**
+    * If the request is OK, it's already been completed and we don't need to wait on it.
+    * The request will be a nullptr, however, so we need to create a new request
+    * and set it to completed to make the "waitall()" function work properly.
+    */
+  } else {
+    ucp_request = (struct ucx_context *)malloc(sizeof(struct ucx_context));
+    ucp_request->completed = 1;
+    ucp_request->needs_release = false;
+  }
+
+  return ucp_request;
+}
+
+/**
+ * @bried Asynchronously receive data from given endpoint with the given tag.
+ */
+struct ucx_context *ucp_irecv(ucp_worker_h worker, ucp_ep_h ep_ptr, void *buf,
+                              int size, int tag) {
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
+
+  struct ucx_context *ucp_request = (struct ucx_context *)ucp_tag_recv_nb(
+    worker, buf, size, ucp_dt_make_contig(1), ucp_tag, default_tag_mask,
+    recv_handle);
+
+  /**
+   * If error, endpoint is closed.
+   * @TODO: Should the request be closed here as well? How should we handle this?
+   * @TODO: Should the Python layer be attempting to reconnect endpoints when they are closed?
+   */
+  if (UCS_PTR_IS_ERR(ucp_request)) {
+    printf("unable to receive UCX data message (%d)\n");
```

Comment:
Better use an assert?

Reply:
I had held off on throwing an exception in these cases because both the original `hello_world.c` UCX example, as well as `ucx-py`, just print to stderr.

Reply:
I will add ASSERT here and gather Akshay's thoughts as well.

Reply:
I think for an example printing to stderr is fine. However for a library like cuML, especially if called from ipython notebooks such errors can be easily burried at places nobody sees them. And maybe even more importantly if an application calls into cuML it can be reallly annyoing if the library prints stuff that the caller has no control over.